### ISSUE TITLE
use of scoreDef and staffDef between measures in CMN

### ISSUE THREAD

**Issue by kepper** [music-encoding/music-encoding#324]
As discussed in https://github.com/rism-ch/verovio/issues/238, the current situation for using `<scoreDef>` and / or `<staffDef>` between measures is somewhat unclear and may need clarification. 
My suggestion is to always require `<scoreDef>` outside of `<measure>`, and only use `<staffDef>` inside `<measure>`. However, this results in the need for an intermediate `<staffGrp>` in the first case. @pe-ro therefore argues that this causes problems: 

> Requiring the use of `<staffGrp>` to effect changes to a single staff creates a problem -- what's supposed to happen to the staves not mentioned in the new `<staffGrp>`? For me, it makes sense to use `<staffGrp>` to mark changes in the number, order and grouping of staves and `<staffDef>` to signal changes in the individual staves.

It's actually true that the grouping function is the sore spot in this, and admittedly, I have no final solution for that yet. However, I'd like to disagree with @pe-ro in some points. The presence or absence of a `<staffDef>` doesn't make any statement whether this staff is rendered or not in a "compressed" full score: That's the very purpose of `@visible` on `<staffDef>`. Everytime a `<staffDef>` isn't included in a later `<scoreDef>`, it preserves all the values specified earlier for this particular `@n` (which may actually be a compilation taken from multiple `<staffDef>` elements). 

If I want to change the meter somewhere in the piece, all I need to do is to add a single `<scoreDef>` with the appropriate attributes. Since they inherit from there, I don't need to specify individual `<staffDef>` elements (inside the mandatory `<staffGrp>`). However, the absence of these `<staffDef>`s can't possibly mean that the whole score becomes invisible!

If it's true that we only need to specify the `<staffDef>`s that actually require some new statement, while all the others are "mentally preserved", then we need to clarify how grouping of staves is affected by this. At least I have used only the first, mandatory `<staffGrp>` in subsequent `<scoreDef>`s. I actually never had to do this so far, but if I had to process this, I would try to make best use of the groups specified in the very first `<scoreDef>` to render the now visible staves. I would need to see some more examples to decide whether this is feasible or not. Maybe @lpugin has more experience on this? 

In out discussion, we just need to make sure we distinguish two cases here: a) the number of (rendered) staves changes, and b) some attributes on a subset of staves are changed, so that all other `<staffDef>`s are just omitted for convenience. Our solution must be perfectly clear for both cases.

**Comment by pe-ro** [comment]
> In [our] discussion, we just need to make sure we distinguish two cases here: a) the number of (rendered) staves changes, and b) some attributes on a subset of staves are changed, so that all other <staffDef>s are just omitted for convenience. Our solution must be perfectly clear for both cases.

@kepper --

My suggestion is to use `<scoreDef>` when the number of staves changes, and `<staffDef>` for changing a subset of staves. That is, 
```xml
<scoreDef>
  <staffDef n="1" ... />
  <staffDef n="2" ... />
  <staffDef n="3" ... />
</scoreDef>
```
regardless of where it occurs within a document, sets the number of rendered staves to 3 and provides attributes for each staff, while
```xml
<staffDef n="2" ... />
<staffDef n="3" ... />
```
as children of `<section>`, allows one to change the attributes for staves 2 and 3 only.

Because `<scoreDef>` resets the entire score layout, one must specify all staves (rendered or not) existing at that point in the score.  If this "score-level reset" is not what's desired, then a `<staffDef>` element should be used for each staff needing a change to its attributes.  Using this approach eliminates any confusion around whether staves missing from `<scoreDef>` are "carried over"; that is, they never are.

**Comment by lpugin** [comment]
@pe-ro : do I understand correctly that an empty `scoreDef` can by used to change one attribute but on all staves? for example

```xml
<scoreDef meter.sym="common"/>
```

meaning that the number / order of staves remain the same?

**Comment by pe-ro** [comment]
Allow me to amend my last comment -- `<scoreDef>` can be used to reset the attributes of the score without specifying those of the staves; that is, in the case of 
```xml
<scoreDef spacing.system="435"/>
```
**all** the previously existing staves are "carried over", since a score without staves isn't much of a score. :-)

**Comment by pe-ro** [comment]
Yes.

**Comment by lpugin** [comment]
OK. Then with a score with transposing instruments, changing a key signature can look like

```xml
<scoreDef key.sig="4s"/>
<!-- changing the transposing instruments -->
<staffDef n="2" key.sig="6s"/>
<staffDef n="3" key.sig="6s"/>
```

**Comment by pe-ro** [comment]
Just to be clear -- there are files in the sample set, converted from MusicXML, that don't follow the plan I've outlined here, Joplin's Elite Syncopations, for example.  At some point, the musicxml2mei converter will need to be updated.

Also, the RNG schema can't enforce these expectations.  We may arrive at some helpful Schematron rules later.

**Comment by pe-ro** [comment]
@lpugin, Yes, this is the only way to preserve all the attributes (from previous `<scoreDef>` and `<staffDef>` elements) and override only the ones you want to change.

**Comment by bwbohl** [comment]
Just to clarify:
considering I want a staff not being rendered from a certain measure onward I could add

```xml
<staffDef n="1" visible="false"/>
```

at an arbitrary level (of course in the right "temporal" spot) in the structure, i.e. `<section>`, `<measure>`, `<layer>` (and some more that I ignore for the moment)?
So `<staff n="1">` could contain it's own visibility change?!

**Comment by lpugin** [comment]
In `layer`? Does not make much sense to me...

**Comment by pe-ro** [comment]
@bwbohl --

> So `<staff n="1">` could contain it's own visibility change?!

I assume your use of the exclamation mark is to register your surprise that a staff could relate information about itself.  In fact, this happens all the time.  For example, a change of clef is a signal that something is about to change with regard to the staff.  So, if a staff can say "I'm using a treble clef now", why shouldn't it also be able to say "I'm invisible now"?

@lpugin --

It doesn't really matter where (within the `<staff>` markup) the change occurs.  Since `<layer>` is a child of `<staff>`, the change can be recorded there just as well as in `<staff>`.

**Comment by lpugin** [comment]
> It doesn't really matter where (within the <staff> markup) the change occurs. Since <layer> is a child of <staff>, the change can be recorded there just as well as in <staff>

I am not sure I can visualise what it means to say that somewhere in a layer the staff to which it belongs is invisible, but as long as you do it does not matter indeed

**Comment by pe-ro** [comment]
😀

**Comment by pe-ro** [comment]
I'm not sure I can visualize that circumstance either, but that doesn't mean we should disallow it just because we can't yet see the usefulness of it.

Would it totally freak you out to contemplate that a layer can contain changes to the score? 😀

**Comment by bwbohl** [comment]
Of course a staff can contain info about itself. I was just picking the example of switching visibility as it is the most bewildering ;-)
There are cases in modern scores where a staff stops being rendered somewhere in the "middle" of a measure, in the middle of a page and its space kept blank just to the next page turn, then on the next page it would be omitted completely. Can we cover such things with the model proposed?

**Comment by kepper** [comment]
The observed problem is still there (we guess), but it might be solved differently. First, impact on different repertoires should be considered. Then, it might be interesting to think about the use of clef etc. elements inside layer. The other thing to keep in mind is changing keys at scoreDef, which gets complicated for transposing instruments
